### ISSUE TITLE
Trigger app uninstall via API?

### ISSUE THREAD

**Issue by clement911** [stripe/stripe-apps#837]
How can we trigger uninstallation of our app via the API?

For example, we might want to uninstall our app from a certain account if their free trial has expired or if we deem that their account is fraudulent. This is important because we don't want to keep receiving webhooks for these accounts for ever.

As a Connect Extension, we could call the following endpoint, but I don't think it applies to Stripe apps:
https://stripe.com/docs/connect/oauth-reference#post-deauthorize

**Comment by clement911** [comment]
I was thinking about a bit more about the flow when the trial for our app expires.

When the trial end date has past, ideally, we could instruct Stripe to stop delivering webhooks, while keeping the app installed so that the user can still open our app's drawer and read a message asking them to enter their payment details.

**Comment by clement911** [comment]
Is there any updates on this?

**Comment by auchenberg-stripe** [comment]
Hi @clement911,

The mental model for Stripe Apps and the installation process is very similar to iOS and Android, where the end-user is in full control of what is being installed into their Stripe Accounts.

Users have to explicitly grant permissions upon installing apps, and they are also in full control on deciding when apps get uninstalled from their accounts.

For this reason, you as a app developer, won't be able to in trigger uninstallation on behalf of your users. Instead you will be notified via webhooks if a user chooses to uninstall your app.

If a user has a active or expired subscription within your product, it's up to you to manage how you want to deal with such scenarios.

For your scenario of a expired trial. What primitives would you ideally need beyond uninstall?

**Comment by clement911** [comment]
Hi @auchenberg-stripe 

We had another chat internally about this and we believe this is a big problem.

The main issue relates to the delivery of undesired webhook events.

As you know, delivery of webhooks can incur serious server costs because they can generate some serious load, especially for large accounts.

Under the current model, if an extremely large stripe account installs our app, Stripe will continue delivering a ton of webhook events until they decide (if ever) to uninstall our app. This will happen regardless of whether they use our app and/or convert to a paying account. (Note that our app is not free and we incur real infrastructure cost to process all webhook events).

Now imagine that, over time, many stripe accounts do the same thing. That is, they install our app, never convert to paying customers, and never uninstall our app either. Over time, we are slowly being bombarded with ever more webhooks that we don't wish to process. This will ballon our server costs for no good reason.

Let me say again that, as a Connect extension, we did not have this problem because we could trigger an uninstall by using the [revoke access endpoint](https://stripe.com/docs/connect/oauth-reference#post-deauthorize).
We believe we need a similar API to avoid the aforementioned issue.

I should also note that we integrate with several other APIs, such as Shopify, QuickBooks and more. They **all** offer a way to programmatically disconnect.

There is an alternative solution that I can think of.
In the Shopify API, we can decide which webhook events to subscribe to on a per account/store basis.
If Stripe had a similar capability, this could solve the problem.
We could decide on a per-account basis to stop listening to most event types and only listen to the bare minimum event types, such as `account.application.deauthorized`. That seems like a much bigger feature that merely providing a disconnect endpoint though.

**Comment by rfelix2121** [comment]
We very much have a similar issue as well. As a Connect extension (for about 10 years), we have ben seeing **many** connected accounts that are fraudulent, and this has only started happening since Stripe Apps rolled out.

Over the past few months, I've noticed a lot of fraudulent accounts being connected to our Stripe account in this way. We initially thought that this was happening via our website, but recently I realized that this is a side effect of how Apps work. 

Anyone can connect their Stripe account to our App from the Stripe Dashboard side, and then we start receiving webhooks for Stripe accounts that will potentially never become customers. On our website, these accounts are easy enough to block with code, but the way that Apps work from the Stripe Dashboard side opens up an abuse/attack vector that means that we are processing webhooks for accounts that will **never** be customers, and which make it more difficult for us to process webhooks on behalf of our actual paying customers. 

Even Stripe thinks that these accounts are fraudulent, because eventually their status changes to "Rejected", or never moves on from "Enabled" to "Complete". 

Right now, we still have access to the Connect extension endpoint to disconnect these fraudulent accounts, and it appears to work... but it's giving us additional work to do because now we have to keep up with identifying and disconnecting these fraudulent accounts, instead of writing code that benefits our paying customers.

It's unclear to me whether the Connect extension endpoint will continue to work in the future (the Apps vs Connect extensions story is a bit confusing to me, since they somewhat seem to be built on top of some of the existing Connect infrastructure), but if it is scheduled to go away at some point, we definitely need a replacement for it, otherwise it will have a material impact on our business.

It doesn't matter whether our App is installed in the user's Stripe account, only that we can turn off webhooks for them if we determine that they are fraudulent. 

If I disconnect an account using the existing Connect webhook, does that remove the App from their Stripe account? Either way, I assume that if they then came to our website and connected using an Install Link, that would install the App into their Stripe account if it wasn't already there. Is that correct?

Thanks!

**Comment by clement911** [comment]
Just to add one more thought.

With install links, our users can now connect their accounts directly from our own app/site.
For symmetry and consistency, I think it's natural for users to expect that they should be able to disconnect it as well.

**Comment by firstpromoter** [comment]
Is there any update regarding this issue?

We are facing the same @clement911 and @rfelix2121, tons of users in trial or who have cancelled the account that don't bother to uninstall / disconnect the app. This is not only of getting useless webhooks that increase server costs but it's an extreme issue related to data privacy, you are getting private data from someone that is not your customer anymore.

**Comment by clement911** [comment]
We haven't found a work around and don't believe there are any.

**Comment by clement911** [comment]
@auchenberg-stripe any updates on this?
Imagine that the largest Stripe merchant installs our app (Maybe Uber or Shopify).
We would instantly start receiving millions of webhooks per second, regardless of whether they become a paying customer. Forever!
This could potentially cripple any app.
What do you suggest here?

**Comment by crysper** [comment]
@clement911 Agree, that's a big issue and i don't know how their engineers missed this. If you're using webhooks on the app, you need to be prepared to have your servers takend down if some big user installs your app even by mistake and you can't do anything from your side...

**Comment by sca1235** [comment]
Having same issue but wondering if we can just manually remove them from our connected account.

does this stop the webhooks? I assume it would and be a fix.

![image](https://github.com/user-attachments/assets/8212b505-e395-430a-871b-1e61a73736bc)

**Comment by Suzan-Dev** [comment]
is this still not implemented?
